### ISSUE TITLE
For embedded relationship STIX SROs `created` and `modified` properties are not being assigned correctly

### ISSUE THREAD

**Issue by himynamesdave** [muchdogesec/stix2arango#16]
This appears to have occurred since: https://github.com/muchdogesec/stix2arango/pull/14

However, this cannot be 100% validated.

Running

```
python3 -m unittest tests/test_18-testing-nested-embedded-ref.py
```

See the `modified` and `created` times in `_is_ref` objects is `null`

```
FOR doc IN test18_edge_collection
                FILTER doc._is_ref == true
                AND doc._is_latest == true
                LET keys = ATTRIBUTES(doc)
                LET filteredKeys = keys[* FILTER !STARTS_WITH(CURRENT, "_")]
                RETURN KEEP(doc, filteredKeys)
```

```
[
  {
    "created": null,
    "created_by_ref": "identity--72e906ce-ca1b-5d73-adcd-9ea9eb66a1b4",
    "id": "relationship--54d79501-6515-581f-b1a4-001d12c07560",
    "modified": null,
    "object_marking_refs": [
      "marking-definition--904ac99b-7539-5de7-9ffa-23186f0e07b6",
      "marking-definition--27557362-b745-4161-96e8-ccd62ce4cb26",
      "marking-definition--94868c89-83c2-464b-929b-a1a8aa3c8487"
    ],
    "relationship_type": "address",
    "source_ref": "cryptocurrency-transaction--f437c493-b651-5cbb-845a-3dd231a39ec6",
    "spec_version": "2.1",
    "target_ref": "cryptocurrency-wallet--d3efcf4e-d9fa-578d-a83f-f11af4c2504c",
    "type": "relationship"
  },
  {
    "created": null,
    "created_by_ref": "identity--72e906ce-ca1b-5d73-adcd-9ea9eb66a1b4",
    "id": "relationship--bec337cc-0433-53a8-bd35-c44e1db4f3c4",
    "modified": null,
    "object_marking_refs": [
      "marking-definition--904ac99b-7539-5de7-9ffa-23186f0e07b6",
      "marking-definition--27557362-b745-4161-96e8-ccd62ce4cb26",
      "marking-definition--94868c89-83c2-464b-929b-a1a8aa3c8487"
    ],
    "relationship_type": "address",
    "source_ref": "cryptocurrency-transaction--f437c493-b651-5cbb-845a-3dd231a39ec6",
    "spec_version": "2.1",
    "target_ref": "cryptocurrency-wallet--f17481a9-a6ad-5ebc-af2a-9b637dab774b",
    "type": "relationship"
  },
  {
    "created": null,
    "created_by_ref": "identity--72e906ce-ca1b-5d73-adcd-9ea9eb66a1b4",
    "id": "relationship--5d606a3b-7b13-5404-8ce2-59438323a64a",
    "modified": null,
    "object_marking_refs": [
      "marking-definition--904ac99b-7539-5de7-9ffa-23186f0e07b6",
      "marking-definition--27557362-b745-4161-96e8-ccd62ce4cb26",
      "marking-definition--94868c89-83c2-464b-929b-a1a8aa3c8487"
    ],
    "relationship_type": "address",
    "source_ref": "cryptocurrency-transaction--f437c493-b651-5cbb-845a-3dd231a39ec6",
    "spec_version": "2.1",
    "target_ref": "cryptocurrency-wallet--9edba811-a14c-5795-8dd6-e39b4bde1c40",
    "type": "relationship"
  },
  {
    "created": null,
    "created_by_ref": "identity--72e906ce-ca1b-5d73-adcd-9ea9eb66a1b4",
    "id": "relationship--5c52661b-b0ea-5387-907f-eae2e6c0d548",
    "modified": null,
    "object_marking_refs": [
      "marking-definition--904ac99b-7539-5de7-9ffa-23186f0e07b6",
      "marking-definition--27557362-b745-4161-96e8-ccd62ce4cb26",
      "marking-definition--94868c89-83c2-464b-929b-a1a8aa3c8487"
    ],
    "relationship_type": "address",
    "source_ref": "cryptocurrency-transaction--f437c493-b651-5cbb-845a-3dd231a39ec6",
    "spec_version": "2.1",
    "target_ref": "cryptocurrency-wallet--7052df7c-e9f4-5fc9-bc57-304278647176",
    "type": "relationship"
  },
  {
    "created": null,
    "created_by_ref": "identity--72e906ce-ca1b-5d73-adcd-9ea9eb66a1b4",
    "id": "relationship--35f9c60e-5364-556e-a6f0-ccb0179eec02",
    "modified": null,
    "object_marking_refs": [
      "marking-definition--904ac99b-7539-5de7-9ffa-23186f0e07b6",
      "marking-definition--27557362-b745-4161-96e8-ccd62ce4cb26",
      "marking-definition--94868c89-83c2-464b-929b-a1a8aa3c8487"
    ],
    "relationship_type": "object-marking",
    "source_ref": "cryptocurrency-transaction--f437c493-b651-5cbb-845a-3dd231a39ec6",
    "spec_version": "2.1",
    "target_ref": "marking-definition--904ac99b-7539-5de7-9ffa-23186f0e07b6",
    "type": "relationship"
  },
  {
    "created": null,
    "created_by_ref": "identity--72e906ce-ca1b-5d73-adcd-9ea9eb66a1b4",
    "id": "relationship--aba6ef16-1d33-5e08-bcb6-86c0696071f0",
    "modified": null,
    "object_marking_refs": [
      "marking-definition--904ac99b-7539-5de7-9ffa-23186f0e07b6",
      "marking-definition--27557362-b745-4161-96e8-ccd62ce4cb26",
      "marking-definition--94868c89-83c2-464b-929b-a1a8aa3c8487"
    ],
    "relationship_type": "object-marking",
    "source_ref": "cryptocurrency-transaction--f437c493-b651-5cbb-845a-3dd231a39ec6",
    "spec_version": "2.1",
    "target_ref": "marking-definition--27557362-b745-4161-96e8-ccd62ce4cb26",
    "type": "relationship"
  },
  {
    "created": null,
    "created_by_ref": "identity--72e906ce-ca1b-5d73-adcd-9ea9eb66a1b4",
    "id": "relationship--b1faec5b-d89c-58aa-a2d6-0b41414fc47f",
    "modified": null,
    "object_marking_refs": [
      "marking-definition--904ac99b-7539-5de7-9ffa-23186f0e07b6",
      "marking-definition--27557362-b745-4161-96e8-ccd62ce4cb26",
      "marking-definition--94868c89-83c2-464b-929b-a1a8aa3c8487"
    ],
    "relationship_type": "object-marking",
    "source_ref": "cryptocurrency-transaction--f437c493-b651-5cbb-845a-3dd231a39ec6",
    "spec_version": "2.1",
    "target_ref": "marking-definition--94868c89-83c2-464b-929b-a1a8aa3c8487",
    "type": "relationship"
  }
]
```

## EXPECTED

Expected output is 

```json
{
	"type": "relationship",
	"spec_version": "2.1",
    "id": "relationship--<UUID v5>",
	"created_by_ref": "<IMPORTED STIX2ARANGO IDENTITY ID>",
	"relationship_type": "<EMBEDDED RELATIONSHIP TYPE WITHOUT REF OR REFS AND _ REPLACED WITH ->",
	"created": "<OBJECT CREATED TIME>",
	"modified": "<OBJECT MODIFIED TIME>",
    "source_ref": "<OBJECT WITH REF PROPERTY>",
    "target_ref": "<REF TARGET PROPERTY>",
    "object_marking_refs": [
        "<object_marking_refs SHOULD MATCH SOURCE_REF OBJECT>"
    ]
}
```

Note, `created` is first time this relationship `id` was generated, and should never be changed

On first creation, `created` and `modified` times are the same. If any updates to this embedded relationship happen (`id`) then the new object has the same `created` time, but an increased modified time.

## ANALYSIS

Currently embedded SROs try to hook into the created and modified time of the source_ref object

This approach causes failures b/c SROs (e.g. ipv4-addr, crytocurrency-wallet, etc) do not have created and modified properties.

Hence you need to use the server creation time, but also have the logic to keep the `created` time for the `id` persistent when updates happen to original objects that contain the embedded refs these properties belong to.

## TESTS

I've added some tests for this

https://github.com/muchdogesec/stix2arango/commit/e2088134e5a21651d04a34b3809d27119e3be987

**Comment by himynamesdave** [comment]
@fqrious I noticed the docs for this were wrong. I have updated to cover this case.

**Comment by fqrious** [comment]
This is kinda hard to implement, what do you think about using a default created time and a modified time instead? if we have the default creation time as 2020-01-01, there won't be a need to query the database for existing date first.

```
created: 2020-01-01
modified: <current-time>
```

**Comment by himynamesdave** [comment]
@fqrious we won't be able to get away with that.

`created` time is important.

If this logic is complex, can't we just implement the same logic that already exists for updating SDOs (that is other objects with `created` and `modified` times? That is very similar to this approach, and would achieve the same thing?

**Comment by fqrious** [comment]
No, existing SDOs come with their own created time, we don't need to probe for old upload's created time for anything

**Comment by himynamesdave** [comment]
To try and demo what I mean...

### background 

See the update behaviour of entries with no modified time

```
python3 -m unittest tests/test_10-updating-object-no-modified-time.py
```

Here is updated object

```
FOR doc IN test10_vertex_collection
            FILTER doc.id == "software--55388d12-8d7d-5ed1-b324-817a293a6854"
            RETURN [doc]
```

see no modified time

```
[
  [
    {
      "_key": "software--55388d12-8d7d-5ed1-b324-817a293a6854+2024-07-02T07:24:52.132979",
      "_id": "test10_vertex_collection/software--55388d12-8d7d-5ed1-b324-817a293a6854+2024-07-02T07:24:52.132979",
      "_rev": "_iFGcGO6---",
      "type": "software",
      "spec_version": "2.1",
      "id": "software--55388d12-8d7d-5ed1-b324-817a293a6854",
      "name": "sha2 Project sha2 0.4.1 for Rust",
      "cpe": "cpe:2.3:a:sha2_project:sha2:0.4.1:*:*:*:*:rust:*:*",
      "swid": "8E3664E0-50A8-4AC3-8863-5552DD536F2F",
      "languages": [
        "en"
      ],
      "vendor": "sha2_project",
      "version": "0.4.1",
      "object_marking_refs": [
        "marking-definition--613f2e26-407d-48c7-9eca-b8e91df99dc9",
        "marking-definition--de23ef3b-83bf-56a9-95c3-46bc1703966c"
      ],
      "_bundle_id": "bundle--bc78b4dd-fba0-4382-9447-e9f3c92b0629",
      "_file_name": "sco-original.json",
      "_stix2arango_note": "test10",
      "_record_md5_hash": "054a3a6db403ecb5a79c4635e0e122c1",
      "_is_latest": false,
      "_record_created": "2024-07-02T07:24:52.132979",
      "_record_modified": "2024-07-02T07:24:52.132979"
    }
  ],
  [
    {
      "_key": "software--55388d12-8d7d-5ed1-b324-817a293a6854+2024-07-02T07:24:53.159098",
      "_id": "test10_vertex_collection/software--55388d12-8d7d-5ed1-b324-817a293a6854+2024-07-02T07:24:53.159098",
      "_rev": "_iFGcGO6--_",
      "type": "software",
      "spec_version": "2.1",
      "id": "software--55388d12-8d7d-5ed1-b324-817a293a6854",
      "name": "New name",
      "cpe": "cpe:2.3:a:sha2_project:sha2:0.4.1:*:*:*:*:rust:*:*",
      "swid": "8E3664E0-50A8-4AC3-8863-5552DD536F2F",
      "languages": [
        "en"
      ],
      "vendor": "sha2_project",
      "version": "0.4.1",
      "object_marking_refs": [
        "marking-definition--613f2e26-407d-48c7-9eca-b8e91df99dc9",
        "marking-definition--de23ef3b-83bf-56a9-95c3-46bc1703966c"
      ],
      "_bundle_id": "bundle--956a753b-5dab-4a45-8b9e-23243fb928d0",
      "_file_name": "sco-updated.json",
      "_stix2arango_note": "test10",
      "_record_md5_hash": "bfe735d6cb1d75cfda854559db90022b",
      "_is_latest": true,
      "_record_created": "2024-07-02T07:24:53.159098",
      "_record_modified": "2024-07-02T07:24:53.159098"
    }
  ]
]
```

See record modified / created exist time exists

```
FOR doc IN test10_vertex_collection
            FILTER doc.id == "software--55388d12-8d7d-5ed1-b324-817a293a6854"
            SORT doc._is_latest DESC
            RETURN {
                id: doc.id,
                _is_latest: doc._is_latest,
                _record_modified: doc._record_modified,
                _record_created: doc._record_created,
            }
```

```
[
  {
    "id": "software--55388d12-8d7d-5ed1-b324-817a293a6854",
    "_is_latest": true,
    "_record_modified": "2024-07-02T07:24:53.159098",
    "_record_created": "2024-07-02T07:24:53.159098"
  },
  {
    "id": "software--55388d12-8d7d-5ed1-b324-817a293a6854",
    "_is_latest": false,
    "_record_modified": "2024-07-02T07:24:52.132979",
    "_record_created": "2024-07-02T07:24:52.132979"
  }
]
```

### For this ticket

Currently all relationships objects created by stix2arango get their `created` and `modified` times from the source object.

This is the cause of this issue, because not all objects have this time.

So this means for any relationships being generated with a stix object with no modified or created time as the source_ref, the relationship objects should NOT have either a created or modified time (the properties should not be in the objects).

Note, some object only contain a `created` time. In this case, the relationships created from it should inherit a created time, but no modified time

In short, any relationships (inc embedded relationships) created by stix2arango should have created and modified properties that match the source object they were created from.

**Comment by fqrious** [comment]
Oh, you're saying it shouldn't be null... The key should just not be on the object?

That it?

**Comment by himynamesdave** [comment]
Yes, as simple as that
